Mark feeds as visited immediatley (#17)

evilpie opened the pull request.

nt1m commented on background/background.js:
```diff
@@ -140,6 +157,7 @@ const LivemarkUpdater = {
     for (; i < max; i++) {
       const item = items[i];
       const url = item.url || feed.feedUrl;
+      this.itemURLsToFeeds.set(url, folder.id);
```
Is it possible to avoid populating this map if the feature is disabled ?

nt1m commented:
@evilpie Something that could help with memory usage would be not storing urls that already been visited. What do you think?

evilpie commented on background/background.js:
```diff
@@ -140,6 +157,7 @@ const LivemarkUpdater = {
     for (; i < max; i++) {
       const item = items[i];
       const url = item.url || feed.feedUrl;
+      this.itemURLsToFeeds.set(url, folder.id);
```
We could do that, but this way we at least reduce memory usage a little bit? I don't know how likely it is to have two feeds that contain the same link.

evilpie commented:
That seems like a good idea, one idea that I had was to use the origin instead, assuming most feed links are actually different pages at the same origin.

evilpie commented:
On the other hand with "origin" it seems likely that we would refresh too often. I think we should add a separate function that just marks livebookmarks as read/unread without actually updating/fetching or doing anything else.

evilpie commented on pages/settings/settings.html:
```diff
@@ -30,7 +30,7 @@ <h3>Default folder</h3>
+      <div>
         <h3>Bookmark prefixes</h3>
         <p>The supplied strings will be prefixed to bookmarks to indicate status. Note that titles won't update until the next feed refresh.</p>
         <label>Unread:</label>
```
We should probably also change the name of the setting, otherwise existing users won't get this feature automatically.

evilpie commented:
I now introduced hashing to avoid keeping the URL strings alive. I also changed the the map to `Int32 -> String | Array [String]` to handle hash collusion, but also multiple livemarks with the same URL.

There is only one problem left: When to clear the `itemURLHashToFeeds` Map. The old code I had for this was not valid, because without `forceUpdate`, we might never reach the code that actually updates the map.

evilpie commented:
@nt1m I hope you have some idea for clearing the Map.

nt1m commented:
@evilpie Can we do `this.itemURLHashToFeeds.delete(hash);` on historyOnVisited ? It wouldn't cover the case of a feed getting removed or one or two URLs getting replaced, but this should happen infrequently enough to be acceptable. 

Alternatively, since we have a HashMap, feeds with the same URLs would have the same hash prefix right ? Can't we use that fact to remove the relevant entries when a feed is forceUpdated with new entries replacing old ones ?

evilpie commented on background/background.js:
```diff
@@ -31,12 +31,12 @@ const LivemarkUpdater = {
 
     // We use this Map to mark feed items as visited.
-    // Map [int32 -> String | Array [String]]
+    // Map [int32 -> Set[String]]
```
:slightly_frowning_face: I did this intentionally like this to avoid the overhead of another object

evilpie commented:
Okay this should be enough. The only case where we now leak hash entries is when a livemark is deleted.

Should we release a test version or something like that?

nt1m commented:
@evilpie A test version seems like a great idea! Thanks for your work on this!
